### PR TITLE
refactor: timer heap

### PR DESCRIPTION
**Pull request by KaoImin**

### CODE REVIEW

**Comment by leeyr338** [comment]
This PR is related to
https://github.com/cryptape/cita/issues/534

We need to make sure all the changes have fixed the issue.

**Comment by KaoImin** [comment]
> This PR is related to
> [cryptape/cita#534](https://github.com/cryptape/cita/issues/534)
> 
> We need to make sure all the changes have fixed the issue.

I do not think #534 caused by this.

**Comment by leeyr338** [comment]
> > This PR is related to
> > [cryptape/cita#534](https://github.com/cryptape/cita/issues/534)
> > We need to make sure all the changes have fixed the issue.
> 
> I do not think #534 caused by this.

But we must point out why this bug always (100%) `panic` at `cita_bft::core::votetime::WaitTimer::start:`: 
```
2019-04-29T16:31:20.806725735+08:00 - INFO - address[0xef3aed75587ea544b9d8df507b42e29baeded7d1] is not consensus power !
2019-04-29T16:31:20.816863399+08:00 - ERROR -
============================
stack backtrace:
   0:     0x56145b125a4d - backtrace::backtrace::trace::hb70975eebea9c40e
   1:     0x56145b125632 - backtrace::capture::Backtrace::new::h9f9ed123ff1e30ee
   2:     0x56145aecb284 - panic_hook::panic_hook::h2e58e0f6c6bffba5
   3:     0x56145aecafb8 - core::ops::function::Fn::call::hb9f6eae8621c8e50
   4:     0x56145b1c46c8 - rust_panic_with_hook
                        at src/libstd/panicking.rs:482
   5:     0x56145b1c4161 - continue_panic_fmt
                        at src/libstd/panicking.rs:385
   6:     0x56145b1c4045 - rust_begin_unwind
   7:     0x56145b1dd8dc - panic_fmt
                        at src/libcore/panicking.rs:85
   8:     0x56145b1dd81b - panic
                        at src/libcore/panicking.rs:49
   9:     0x56145ae2a2c2 - cita_bft::core::votetime::WaitTimer::start::hd8b90ccaa5f178ec
  10:     0x56145ae318cc - std::sys_common::backtrace::__rust_begin_short_backtrace::h9ec9f7185c88019d
  11:     0x56145ae32fbb - std::panicking::try::do_call::hc8730e50693ed203
  12:     0x56145b1ca049 - __rust_maybe_catch_panic
                        at src/libpanic_unwind/lib.rs:87
  13:     0x56145ae35e6f - <F as alloc::boxed::FnBox<A>>::call_box::hd81aa90afe319fe7
  14:     0x56145b1c93bd - call_once<(),()>
                        at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/liballoc/boxed.rs:759
                         - start_thread
                        at src/libstd/sys_common/thread.rs:14
                         - thread_start
                        at src/libstd/sys/unix/thread.rs:81
  15:     0x7f6c472e36da - start_thread
  16:     0x7f6c46df488e - __clone
  17:                0x0 - <unknown>

position:
Thread <unnamed> panicked at called `Option::unwrap()` on a `None` value, src/libcore/option.rs:345

This is a bug. Please report it at:

    https://github.com/cryptape/cita/issues/new?labels=bug&template=bug_report.md
============================
```

File: src/core/votetime.rs [hunk 1]
```diff
@@ -18,17 +18,29 @@
...
-#[derive(Clone)]
+#[derive(Clone, Eq, PartialEq)]
 pub struct TimeoutInfo {
     pub timeval: Instant,
     pub height: usize,
     pub round: usize,
     pub step: Step,
 }
 
+impl ::std::cmp::PartialOrd for TimeoutInfo {
```
Comment (kaikai1024) [inline src/core/votetime.rs:32]: Where is it used?

Comment (KaoImin) [reply src/core/votetime.rs:32]: Overload compare for `TimeoutInfo` so that the struct can be compared in the heap.

**Comment by leeyr338** [comment]
call stack for the `panic`:
WechatIMG3
WechatIMG4

**Comment by KaoImin** [comment]
> call stack for the `panic`:
> WechatIMG3
> WechatIMG4

If this, the PR will fix the issue #534 in CITA.

File: src/core/votetime.rs (hunk 1 above)
Comment (kaikai1024) [reply src/core/votetime.rs:32]: 下面不是比较的 结构体里的 `timeval` 吗？

Comment (KaoImin) [reply src/core/votetime.rs:32]: > 下面不是比较的 结构体里的 `timeval` 吗？

就是对 `TimeoutInfo` 结构重载比较运算符，这样才能把这个结构放到堆里面，运算符重载的实现就是比较`timeval`。

Comment (kaikai1024) [reply src/core/votetime.rs:32]: 我原来以为是和 `now` 比较。所以是 `peek_min` 的时候用的吧

**Review by kaikai1024 — APPROVED** [review]

File: src/core/votetime.rs (hunk 1 above)
Comment (KaoImin) [reply src/core/votetime.rs:32]: 能放入堆中的元素一定是能够比较大小的，不管是大顶堆还是小顶堆。原来没有重载比较运算符，在堆中只插入 timeval，另外存放一个 Hashmap 解决的

**Comment by kaikai1024** [comment]
> If this, the PR will fix the issue #534 in CITA.

How?

**Pull request merged**